src/Components/Web/src/Routing/NavigationLock.cs

```diff
@@ -0,0 +1,89 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using Microsoft.JSInterop;
+
+namespace Microsoft.AspNetCore.Components.Routing;
+
+/// <summary>
+/// A component that can be used to intercept navigation events. 
+/// </summary>
+public class NavigationLock : IComponent, IAsyncDisposable
+{
+    private bool HasOnBeforeInternalNavigationCallback => OnBeforeInternalNavigation.HasDelegate;
+
+    [Inject]
+    private IJSRuntime JSRuntime { get; set; } = default!;
+
+    [Inject]
+    private NavigationManager NavigationManager { get; set; } = default!;
+
+    /// <summary>
+    /// Gets or sets a callback to be invoked when an internal navigation event occurs.
+    /// </summary>
+    [Parameter]
+    public EventCallback<LocationChangingContext> OnBeforeInternalNavigation { get; set; }
+
+    /// <summary>
+    /// Gets or sets whether a browser dialog should prompt the user to either confirm or cancel
+    /// external navigations.
+    /// </summary>
+    [Parameter]
+    public bool ConfirmExternalNavigation { get; set; }
+
+    void IComponent.Attach(RenderHandle renderHandle)
+    {
+    }
+
+    async Task IComponent.SetParametersAsync(ParameterView parameters)
```

Comment:
The only reason I did this was to avoid adding these methods to the public API. Do you think that's a valid concern or is it fine to change it?

Reply:
In most cases for equivalent components we did regular interface implementations (e.g., `RouteView`). However I don't know of any reason why it would be a problem to do it like you have. Either way is fine with me.

Reply:
I would follow the established pattern unless there is an explicit concrete benefit of using an alternative.